rfcs/004-typechecking.md

```diff
@@ -0,0 +1,496 @@
+deemed impredicative polymorphism to be more important and chose to abandon the
+variance of arrows and other type constructors. Coupled with a so-called
+QuickLook phase, Haskell is able to most common impredicative instantiations.
+
+We can't apply the Haskell approach naively, because in the presence of `Dyn`,
+an impredicative instantiation is not the only possibility anymore, even if we
+don't do deep instantiation or deep skolemization:
+
+```nickel
+let foo : Dyn = null in
+let id : forall a. a -> a = fun x => x in
+let concat : forall b. Array b -> Array b -> Array b = (@) in
+
+concat [id] [null]
+```
+
+In the QuickLook setting, when we see the argument `[id]`, we know that `b` must
+be instantiated with `forall a. a -> a`. This is not true here, as `b` could
+also be `Dyn` (and in this case, given the second argument, it should!). We
+can't delay solving constraints involving polymorphic function types as we could
+do with monomorphic types (in that case, just generating `?a >: forall a. a ->
+a` at the first argument without solving right away): when applying a function
+with type `?a`, we need to know if `?a` must be instantiated with an arrow or if
+it has potential heading `forall`s which need to be instantiated.
+
+We may extend the QuickLook phase to also cover `Dyn`, but this needs more
+investigation. Another fix could be to make all type constructors invariant also
+for the subtyping relation generated by `Dyn`, but that is quite restrictive in
+practice (several examples in this document would need `Array T <: Array Dyn`,
+to get rid of the casts).
+
+### Packing up
+
+To sum up (**TODO**: better wording):
+
+- We only want to infer `Dyn` because of an explicit type annotation.
+- Polymorphism and subtyping interact in a non trivial way. It is not obvious
+   how to apply QuickLook in the presence of our proposed subtyping relation.
+
+Given that impredicativity looks complex to combine with subtyping, and isn't
+really a strong motivation for a language like Nickel, a reasonable choice is to
+rather follow the path of Dunfield and Krishnaswami. However, the recent uproar
+upon the removing of the subsumption rule for arrows in Haskell[^4] shows that
+such things (variance of arrows with respect to polymorphism-induced subtyping)
+are way easier and safer to add than to remove, from a user experience
+perspective. In consequence, we propose to take a more restrictive approach
+first, which is still open to either pursing one road or the other if there is
+such a need later.
+
+The final proposal is a system that:
+
+- Has subtyping generated by the dynamic type and dictionaries.
+- Support higher-rank polymorphism
+- Is predicative
+- Doesn't do deep skolemization/instantiation, or equivalently, in which type
+  constructors are invariant with respect to subtyping induced by polymorphism
+
+A declarative specification as well as a constraint-solving algorithm are
+provided in the file [type-system.pdf](./type-system.pdf).
```

Comment:
This link is dead. And I don't think that we want to host a pdf in the git repository anyway. We will have to find a place to host the generated pdf, probably on the website?